google_guest_agent/uefi/uefi_windows.go

```diff
@@ -0,0 +1,179 @@
+	// Token identifies the user, the user's groups, and the user's privileges.
+	procOpenProcessToken = advapiDLL.NewProc("OpenProcessToken")
+	// https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-lookupprivilegevaluew
+	// procLookupPrivilegeValueW is used to retrieve the locally unique identifier (LUID)
+	procLookupPrivilegeValueW = advapiDLL.NewProc("LookupPrivilegeValueW")
+	// https://learn.microsoft.com/en-us/windows/win32/api/securitybaseapi/nf-securitybaseapi-adjusttokenprivileges
+	// procAdjustTokenPrivileges is used for enabling the privileges on the access token.
+	procAdjustTokenPrivileges = advapiDLL.NewProc("AdjustTokenPrivileges")
+)
+
+const (
+	// SE_SYSTEM_ENVIRONMENT_NAME is the privilege required to read a firmware environment variable.
+	SE_SYSTEM_ENVIRONMENT_NAME = "SeSystemEnvironmentPrivilege"
+	// PROC_TOKEN_ADJUST_PRIVILEGES is access required to change the specified privileges.
+	PROC_TOKEN_ADJUST_PRIVILEGES = 0x0020
+	// PROC_SE_PRIVILEGE_ENABLED is privilege attribute used with LUID_AND_ATTRIBUTES stating to
+	// enable the specified privilege.
+	PROC_SE_PRIVILEGE_ENABLED = 0x00000002
+)
+
+// https://learn.microsoft.com/en-us/windows/win32/api/ntdef/ns-ntdef-luid
+// LUID is the opaque identifier structure that is guaranteed to be unique on the local machine.
+// It is used to locally represent the privilege name (e.g. SeSystemEnvironmentPrivilege in this case).
+type LUID struct {
+	LowPart  uint32
+	HighPart int32
+}
+
+// https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-luid_and_attributes
+// LUID_AND_ATTRIBUTES structure represents a locally unique identifier (LUID) and its attributes.
+type LUID_AND_ATTRIBUTES struct {
+	LUID       LUID
+	Attributes uint32
+}
+
+// https://learn.microsoft.com/en-us/windows/win32/api/winnt/ns-winnt-token_privileges
+// TOKEN_PRIVILEGES is the structure that contains information about a set of privileges for an access token.
+type TOKEN_PRIVILEGES struct {
+	PrivilegeCount uint32
+	Privileges     [1]LUID_AND_ATTRIBUTES
+}
+
+// ReadVariable reads UEFI variable and returns as byte array.
+// Throws an error if variable is invalid or empty.
+func ReadVariable(v VariableName) (*Variable, error) {
+
+	logger.Debugf("Enabling required %s priviliges for agent process", SE_SYSTEM_ENVIRONMENT_NAME)
+	if err := enablePrivilege(SE_SYSTEM_ENVIRONMENT_NAME); err != nil {
+		return nil, err
+	}
+
+	name := unsafe.Pointer(syscall.StringToUTF16Ptr(v.Name))
+	guid := unsafe.Pointer(syscall.StringToUTF16Ptr("{" + v.GUID + "}"))
+
+	buffer := make([]byte, 1024)
+
+	// This call returns number of bytes written to the output buffer, unused, error
+	size, _, err := procGetFirmwareEnvironmentVariableW.Call(
+		uintptr(name),
+		uintptr(guid),
+		uintptr(unsafe.Pointer(&buffer[0])),
+		uintptr(uint32(len(buffer))),
+	)
+
+	if size == uintptr(0) {
```

Comment:
Is there a condition of read size > 0 and  `error`?

Reply:
no, thing with these calls is that `error` is always non-nil we need to inspect primary return value. Here return value 0 indicates the failure.